src/NCalc.Core/Parser/LogicalExpressionParser.cs

```diff
@@ -114,7 +114,14 @@ private static Parser<LogicalExpression> CreateExpressionParser(CultureInfo cult
 
         var intNumber = Terms.Number<int>(NumberOptions.Integer)
             .AndSkip(Not(OneOf(Terms.Text("."), Terms.Text("E", true))))
-            .Then<LogicalExpression>(static d => new ValueExpression(d));
+            .Then<LogicalExpression>(static (ctx, val) =>
+            {
+                bool useLong = ((LogicalExpressionParserContext)ctx).Options.HasFlag(ExpressionOptions.LongAsDefault);
+                if (useLong)
+                    return new ValueExpression((long)val);
+
+                return new ValueExpression(val);
+            });
```

Comment:
I think it's possible to optimize performance here avoiding a cast by using `Terms.Number<long>` and doing the if at parser level, but maybe I'm wrong.